### PR TITLE
bs - add course number search option to course description search (frontend & backend)

### LINKED ISSUE #20: Add course number option to course description search

### PR DESCRIPTION
**Pull request by BrennaScholte**
In this pull request, I added the course number as a search option to the course description search page. This is similar to the course history search. If you don't input a course number it will just list all the courses (+ their descriptions) for the selected subject area.

Closes #20 

Before:
Screenshot 2023-05-31 at 10 13 32 PM

After:
Screenshot 2023-05-31 at 10 05 52 PM

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -10,8 +10,8 @@ Users with a Google Account can also store past, current or future schedules of
 
 | Type | Link       | 
 |------|------------|
-| prod | <https://proj-courses.dokku-00.cs.ucsb.edu/> | 
-| qa | <https://proj-courses-qa.dokku-00.cs.ucsb.edu/>  | 
+| prod | <https://proj-courses.dokku-12.cs.ucsb.edu/> | 
+| qa | <https://proj-courses-qa.dokku-12.cs.ucsb.edu/>  |
```
Comment (pconrad) [inline README.md:15]: This ought to be in it's own PR for 5 points... if you include it in this one, you are cheating yourself and your team out of 5 points.

File: frontend/src/main/components/BasicCourseSearch/BasicCourseSearchForm.js [hunk 2]
```diff
@@ -37,10 +38,29 @@ const BasicCourseSearchForm = ({ fetchJSON }) => {
...
+    
+    if (rawCourse.match(/[a-zA-Z]+/g) != null) {
+        const suffix = rawCourse.match(/[a-zA-Z]+/g)[0];
+        setCourseSuf(suffix);
+    } else {
+        setCourseSuf("");
+    }
   };
 
   // Stryker disable all : Stryker is testing by changing the padding to 0. But this is simply a visual optimization as it makes it look better
```
Comment (pconrad) [inline frontend/src/main/components/BasicCourseSearch/BasicCourseSearchForm.js:66]: The stryker disable all here is not ok... we can write a test for the CSS, and even if we couldn't disabling all stryker mutation testing for the entire component leaves us vulnerable to lots of testing gaps.

Please add a test for the CSS property, and remove this.

I'm aware that it was in the component before, but we need to address this tech debt.

Comment (pconrad) [reply frontend/src/main/components/BasicCourseSearch/BasicCourseSearchForm.js:66]: Especially since you are changing code in the area that this test exception covers.

File: frontend/src/main/components/BasicCourseSearch/BasicCourseSearchForm.js [hunk 3]
```diff
@@ -73,6 +93,10 @@ const BasicCourseSearchForm = ({ fetchJSON }) => {
             />
           </Col>
         </Row>
+        <Form.Group controlId="BasicSearch.CourseNumber">
+            <Form.Label>Course Number (Try searching '16' or '130A')</Form.Label>
```
Comment (pconrad) [inline frontend/src/main/components/BasicCourseSearch/BasicCourseSearchForm.js:97]: I wonder if it would be better to say: 

```
Try searching for '16' or '130A', or leave blank to get all courses
```

**Review by pconrad — CHANGES_REQUESTED** [review]
A few little things here and there...

**Review by pconrad — CHANGES_REQUESTED** [review]
Modifying BasicCourseSearchForm means you need to check ALL of the pages where that form is used.   It isn't just the CourseDescriptions page; it's also the home page.  

The new field appears there now, but does not work, introducing what will feel like a bug to end users.  This needs to be addressed (there are several possible solutions.)

File: frontend/src/main/components/BasicCourseSearch/BasicCourseSearchForm.js (hunk 2 above)
Comment (BrennaScholte) [reply frontend/src/main/components/BasicCourseSearch/BasicCourseSearchForm.js:66]: I'm a bit confused on how to go about making a test for a CSS property. I believe this stryker disable was in the starter code. Is there an example of a test somewhere else in the starter code that I can base this off of?

Comment (pconrad) [reply frontend/src/main/components/BasicCourseSearch/BasicCourseSearchForm.js:66]: Here's a link on how to do it: 

https://ucsb-cs156.github.io/topics/stryker/stryker_html_attribute_surviving_mutants.html

File: frontend/src/main/components/BasicCourseSearch/CourseOverTimeSearchForm.js [hunk 4]
```diff
@@ -61,7 +61,7 @@ const CourseOverTimeSearchForm = ({ fetchJSON }) => {
     } else {
         setCourseSuf("");
     }
-};
+  };
 
 
   // Stryker disable all : Stryker is testing by changing the padding to 0. But this is simply a visual optimization as it makes it look better
```
Comment (pconrad) [inline frontend/src/main/components/BasicCourseSearch/CourseOverTimeSearchForm.js:67]: This comments needs to be removed; it's too broad an exception, plus we know how to test for this
(see: https://ucsb-cs156.github.io/topics/stryker/stryker_html_attribute_surviving_mutants.html)

**Review by pconrad — CHANGES_REQUESTED** [review]
Still a couple of stryker issues to address

File: frontend/src/main/components/BasicCourseSearch/CourseOverTimeSearchForm.js (hunk 4 above)
Comment (BrennaScholte) [reply frontend/src/main/components/BasicCourseSearch/CourseOverTimeSearchForm.js]: Please see the three most recent commits below. Is this what you were expecting me to add? Or am I still missing something (in regards to the stryker tests)?

**Comment by BrennaScholte** [comment]
> , but does not work, introducing what will feel like a bug to end users. This needs to be addressed (there are several possible solutions.)

@pconrad What would be your recommendation to go about fixing this? So far I've thought about either implementing the course number search option everywhere BasicCourseSearchForm is used or making a new form for CourseDescription.

**Comment by pconrad** [comment]
> > , but does not work, introducing what will feel like a bug to end users. This needs to be addressed (there are several possible solutions.)
> 
> @pconrad What would be your recommendation to go about fixing this? So far I've thought about either implementing the course number search option everywhere BasicCourseSearchForm is used or making a new form for CourseDescription.

I think a new form that's called `CourseSearchWithCourseNumberForm` makes sense.  You might find that initially it's only being used on the CourseDescriptions page, but later it might be incorporated into other pages too.
